server/broadcast.go

```diff
@@ -185,24 +190,27 @@ func NewSessionManager(node *core.LivepeerNode, params *streamParameters, pl cor
 	maxInflight := common.HTTPTimeout.Seconds() / SegLen.Seconds()
 	numOrchs := int(math.Min(poolSize, maxInflight*2))
 	bsm := &BroadcastSessionsManager{
-		mid:            params.mid,
-		sel:            sel,
-		sessMap:        make(map[string]*BroadcastSession),
-		createSessions: func() ([]*BroadcastSession, error) { return selectOrchestrator(node, params, pl, numOrchs) },
-		sessLock:       &sync.Mutex{},
-		numOrchs:       numOrchs,
+		node:     node,
+		params:   params,
+		pl:       pl,
+		sel:      sel,
+		sessMap:  make(map[string]*BroadcastSession),
+		sessLock: &sync.Mutex{},
+		numOrchs: numOrchs,
+		poolSize: int(poolSize),
+		sus:      newSuspender(),
 	}
 	bsm.refreshSessions()
 	return bsm
 }
 
-func selectOrchestrator(n *core.LivepeerNode, params *streamParameters, cpl core.PlaylistManager, count int) ([]*BroadcastSession, error) {
-	if n.OrchestratorPool == nil {
+func (bsm *BroadcastSessionsManager) createSessions() ([]*BroadcastSession, error) {
```

Comment:
I understand that defining this method on `BroadcastSessionsManager` feels cleaner, but from looking at the tests it actually feels like this change makes writing tests for the refresh behavior of `BroadcastSessionsManager` more difficult. Previously, we could just assign a stub implementation of `createSessions()` to simulate different session lists returned during a refresh. Now, we have to modify a `stubDiscovery` instance to simulate different session lists, but we also have to execute all the code in this `createSessions()` method creating additional dependencies that we have to take into consideration as we update/add BroadcastSessionsManager tests related to refresh behavior.

At the moment, this doesn't feel like a worthwhile trade, but I'm happy to hear arguments for why it would be worthwhile (maybe if there are tweaks to the proposed refactor).

Reply:
I agree that it makes writing the tests easier, experienced that first hand, but should we really be monkey patching a method within a single package? I more of consider that to be a last resort. There's a good balance to everything, isn't this one slightly tipping the scales to the wrong side a little bit? I mean looking at it is is essentially just a method for the type. 

Also I bit the bullet on this because my compiler showed an error trying to pass `suspender` as an argument to the function we assign to `createSessions` but now that I fixed my go tooling that seemed to have been a false error.

Reply:
Agreed with Yondon - the changes to the testing flow is a little heavy-handed, so this is actually one instance where it feels like a worthwhile tradeoff. Moreover, there is a lot of additional state being added to the `BroadcastSessionManager` simply for this one function; I believe that was another reason for keeping `selectOrchestrators` outside the BroadcastSessionManager. The less state we have to carry around, the better.

Reply:
> but should we really be monkey patching a method within a single package?

The fact that we can insert a stub function during tests isn't too different from inserting a stub implementation of an interface during tests (consider if `createSessions` was instead an `SessionCreator` interface with a single `CreateSessions` method). 

I think the reason that the current usage of `createSessions` doesn't feel clean is probably because `NewSessionManager` always uses the same function implementation instead of accepting the function implementation as a parameter and because the current function implementation closes around free variables that are passed to `selectOrchestrator` instead of storing them in a struct.

If the test updates mentioned in the rest of this review can be fixed without introducing a lot of additional test cruft then I think this change may be more appealing.

Reply:
dropped those changes , fixed in 2c78fb9a784ea81c42a9bdbd0d91c71ab6fc255b